Tests/MapboxCoreNavigationTests/RouteControllerTests.swift

```diff
@@ -9,538 +9,1224 @@ import MapboxNavigationNative
+        delegate.onDidFailToRerouteWith = { error in
+            XCTAssertEqual(error as! DirectionsError, .unableToRoute)
+            didFailToReroutExpectation.fulfill()
+        }
+
+        routeController.routeProgress.currentLegProgress.userHasArrivedAtWaypoint = true
+        let status = TestNavigationStatusProvider.createNavigationStatus(routeState: .complete)
+        routeController.rerouteAfterArrivalIfNeeded(rawLocation, status: status)
         
-        locationManager.speedMultiplier = 1
-        locationManager.startUpdatingLocation()
-        wait(for: [routeExpectation], timeout: 15)
+        XCTAssertFalse(navigatorSpy.setRoutesCalled)
+        waitForExpectations(timeout: expectationsTimeout)
+    }
+
+    func testRerouteAfterArrivalIfCloseToStep() {
+        let didFailToReroutExpectation = expectation(description: "Did fail to reroute call on delegate")
+        didFailToReroutExpectation.isInverted = true
+        delegate.onDidFailToRerouteWith = { _ in
+            didFailToReroutExpectation.fulfill()
+        }
+
+        let status = TestNavigationStatusProvider.createNavigationStatus(routeState: .complete)
+        let newLocation = CLLocation(coordinate: options.waypoints[0].coordinate)
+        routeController.rerouteAfterArrivalIfNeeded(newLocation, status: status)
+
+        XCTAssertFalse(navigatorSpy.setRoutesCalled)
+        XCTAssertFalse(routingProvider.calculateRoutesCalled)
+        waitForExpectations(timeout: expectationsTimeout)
+    }
+
+    func testRerouteAfterArrival() {
+        let didFailToReroutExpectation = expectation(description: "Did fail to reroute call on delegate")
+        didFailToReroutExpectation.isInverted = true
+        delegate.onDidFailToRerouteWith = { _ in
+            didFailToReroutExpectation.fulfill()
+        }
+
+        let newResponse = IndexedRouteResponse(routeResponse: differentRouteResponse, routeIndex: 0)
+        routingProvider.returnedRoutesResult = .success(newResponse)
+
+        let status = TestNavigationStatusProvider.createNavigationStatus(routeState: .complete)
+        routeController.rerouteAfterArrivalIfNeeded(rawLocation, status: status)
+
+        XCTAssertTrue(navigatorSpy.setRoutesCalled)
+        XCTAssertEqual(routeController.indexedRouteResponse.currentRoute, newResponse.currentRoute)
+        waitForExpectations(timeout: expectationsTimeout)
+    }
+
+    func testUpdateRouteIfFinishedRouting() {
+        let expectation = expectation(description: "Should not call callback")
+        expectation.isInverted = true
+        let newResponse = IndexedRouteResponse(routeResponse: differentRouteResponse, routeIndex: 0)
+        routeController.finishRouting()
+        routeController.updateRoute(with: newResponse, routeOptions: options) { _ in
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: expectationsTimeout)
+    }
+
+    func testUpdateRouteIfFinishedRoutingIfFail() {
```

Comment:
Following tests mention `finishedRouting` but does not call `routeController.finishRouting()`

Reply:
Renamed, there should not be two `If` in method name